protocols/client/client.go

```diff
@@ -410,25 +419,20 @@ func HybridVSockDialer(sock string, timeout time.Duration) (net.Conn, error) {
 			return nil, err
 		}
 
-		// Receive the packet from the connection without removing it from
-		// the receive queue (MSG_PEEK), ensuring the connection is usable.
-		if uc, ok := conn.(*net.UnixConn); ok {
-			file, err := uc.File()
-			if err != nil {
-				conn.Close()
-				return nil, err
-			}
-			eot := make([]byte, 1)
-			n, _, err := unix.Recvfrom(int(file.Fd()), eot, syscall.MSG_PEEK)
-			file.Close()
-			if err != nil || n == 0 {
-				conn.Close()
-				if err == nil {
-					err = io.EOF
-				}
-				return nil, err
-			}
+		// A trivial handshake is included in the host-initiated vsock connection protocol.
+		// It looks like this:
+		// - [host] CONNECT <port><LF>
+		// - [guest/success] OK <assigned_host_port><LF>
+		reader := bufio.NewReader(conn)
+		response, err := reader.ReadString('\n')
+		if err != nil {
+			conn.Close()
+			return nil, err
```

Comment:
Try removing this line, this way the function will return `conn, nil` and the GRPC code will handle dialing again.

Reply:
I'm confused, if reading was failing, we should return error instantly. Then here in func [commonDialer](https://github.com/kata-containers/agent/blob/master/protocols/client/client.go#L330)
```
	go func() {
		for {
			select {
			case <-cancel:
				// canceled or channel closed
				return
			default:
			}

			conn, err := dialFunc()
			if err == nil {
				// Send conn back iff timer is not fired
				// Otherwise there might be no one left reading it
				if t.Stop() {
					ch <- conn
				} else {
					conn.Close()
				}
				return
			}
		}
	}()
```
we could re-connect(`conn, err := dialFunc()`) again.

Reply:
Yes I agree, but this quick re-connection performed by `commonDialer()` tend to get things to fail. Instead, when relying on the backoff strategy from GRPC (which happens when we return `conn, nil`), things are more stable.
Just to clarify, I'm not saying this is a long term solution, but while we look actively for the root cause, relying on the backoff strategy from GRPC gives us a more stable CI for both Firecracker and Cloud-Hypervisor.